src/rust/include/rust/orchard/incremental_sinsemilla_tree.h

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2020 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_RUST_INCLUDE_INCREMENTAL_MERKLE_H
+#define ZCASH_RUST_INCLUDE_INCREMENTAL_MERKLE_H
+
+#include "rust/streams.h"
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define SINSEMILLA_DIGEST_LEN 32U
+
+/// Pointer to an Orchard incremental merkle tree frontier
+struct OrchardMerkleFrontierPtr;
+typedef struct OrchardMerkleFrontierPtr OrchardMerkleFrontierPtr;
+
+// Create an empty Orchard Merkle frontier.
+//
+// Memory allocated to the resulting value must be manually freed.
+OrchardMerkleFrontierPtr* orchard_merkle_frontier_empty();
+
+// Create an empty Orchard Merkle frontier.
+//
+// Memory allocated to the resulting value must be manually freed.
+OrchardMerkleFrontierPtr* orchard_merkle_frontier_empty();
+
+// Clones the given Orchard Merkle frontier and returns
+// a pointer to the newly created tree. Both the original
+// tree's memory and the newly allocated one need to be freed
+// independently.
+OrchardMerkleFrontierPtr* orchard_merkle_frontier_clone(
+        const OrchardMerkleFrontierPtr* tree_ptr
+        );
+
+// Free the memory allocated for the given Orchard Merkle frontier.
+void orchard_merkle_frontier_free(
+        OrchardMerkleFrontierPtr* tree_ptr
+        );
+
+// Parses an Orchard Merkle frontier from a stream. If parsing
+// fails, this will return the null pointer.
+//
+// Memory allocated to the resulting value must be manually freed.
+OrchardMerkleFrontierPtr* orchard_merkle_frontier_parse(
+        void* stream,
+        read_callback_t read_cb
+        );
+
+// Parses an Orchard Merkle frontier from a stream. If parsing
+// fails, this will return the null pointer.
+//
+// Memory allocated to the resulting value must be manually freed.
+bool orchard_merkle_frontier_serialize(
+        const OrchardMerkleFrontierPtr* tree_ptr,
+        void* stream,
+        write_callback_t write_cb
+        );
+
+// For each action in the provided bundle, append its
+// commitment to the frontier.
+bool orchard_merkle_frontier_append_bundle(
+        OrchardMerkleFrontierPtr* tree_ptr,
+        const OrchardBundlePtr* bundle
```

Comment:
Add an import for `rust/orchard.h` to ensure this is in scope. Or make a `rust/orchard/types.h` if we want to disconnect these.